sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,742 @@
+                .getProgress()
+                .getWorkRemaining();
+      }
+
+      // Before processing elements, we don't have a good estimated size of records and offset gap.
+      if (avgOffsetGap != null) {
+        numOfRecords = numOfRecords / (1 + avgOffsetGap.get());
+      }
+      return (avgRecordSize == null ? 1 : avgRecordSize.get()) * numOfRecords;
+    }
+
+    @SplitRestriction
+    public void splitRestriction(
+        @Element Row kafkaSourceDescription,
+        @Restriction OffsetRange offsetRange,
+        OutputReceiver<OffsetRange> receiver)
+        throws Exception {
+      receiver.output(offsetRange);
+    }
+
+    @NewTracker
+    public RestrictionTracker<OffsetRange, Long> restrictionTracker(
+        @Element Row kafkaSourceDescription, @Restriction OffsetRange restriction) {
+      Map<String, Object> updatedConsumerConfig =
+          overrideBootstrapServersConfig(consumerConfig, kafkaSourceDescription);
+      if (restriction.getTo() == Long.MAX_VALUE) {
+        TopicPartition topicPartition =
+            new TopicPartition(
+                kafkaSourceDescription.getString(KafkaSourceDescriptionSchemas.TOPIC),
+                kafkaSourceDescription.getInt32(KafkaSourceDescriptionSchemas.PARTITION));
+        KafkaLatestOffsetEstimator offsetPoller =
+            new KafkaLatestOffsetEstimator(
+                consumerFactoryFn.apply(
+                    KafkaIOUtils.getOffsetConsumerConfig(
+                        "tracker-" + topicPartition, offsetConsumerConfig, updatedConsumerConfig)),
+                topicPartition);
+        return new GrowableOffsetRangeTracker(restriction.getFrom(), offsetPoller);
+      }
+      return new OffsetRangeTracker(restriction);
+    }
+
+    @ProcessElement
+    public ProcessContinuation processElement(
+        @Element Row kafkaSourceDescription,
+        RestrictionTracker<OffsetRange, Long> tracker,
+        WatermarkEstimator watermarkEstimator,
+        OutputReceiver<KafkaRecord<K, V>> receiver) {
+      TopicPartition topicPartition =
+          new TopicPartition(
+              kafkaSourceDescription.getString(KafkaSourceDescriptionSchemas.TOPIC),
+              kafkaSourceDescription.getInt32(KafkaSourceDescriptionSchemas.PARTITION));
+      Map<String, Object> updatedConsumerConfig =
+          overrideBootstrapServersConfig(consumerConfig, kafkaSourceDescription);
+      try (Consumer<byte[], byte[]> consumer = consumerFactoryFn.apply(updatedConsumerConfig)) {
+        consumerSpEL.evaluateAssign(consumer, ImmutableList.of(topicPartition));
+        long startOffset = tracker.currentRestriction().getFrom();
+        long expectedOffset = startOffset;
+        consumer.seek(topicPartition, startOffset);
+        ConsumerRecords<byte[], byte[]> rawRecords = ConsumerRecords.empty();
+
+        try {
+          while (true) {
+            rawRecords = consumer.poll(kafkaPollTimeout.getMillis());
```

Comment:
Do we really need to poll here, can we check if there is anything left and if there is nothing return a process continuation with a resume delay that represents the poll timeout?

Reply:
It seems like there is no difference between poll ->resume and resume -> poll? For example, a bad situation could be there is only one `TopicPartition` and there is always no available records. In this case, both poll ->resume and resume -> poll will poll the records with a time interval of api timeout + residual reschedule time.